README.md
# QB-Core Trespassing Alarms
A system to create polyzones, with a given chance to alert police on trespassing  
Created by Nikki#9699

![screenshot-alert](https://i.gyazo.com/14efab35c06823a4916b8cbde337bd43.png)

# Preview video

https://streamable.com/g9z7ux

## INTRODUCTION

- A resource where polyzones are setup to protect an area
- Specific Jobs can be ignored (Changeable in config.lua) (If left empty it will always trigger)
- Chance for alert can be changed per polyzone (Changeable in config.lua)
- Set cooldown to prevent continous trigger (Changeable in config.lua)

## DEPENDENCIES
**[qb-core] (https://github.com/qbcore-framework/qb-core)** - QBCore  
**[PolyZone] (https://github.com/mkafrin/PolyZone)** - To create the zones  
**[interact-sound] (https://github.com/plunkettscott/interact-sound)** - To play audio files  
**[ps-dispatch] (https://github.com/Project-Sloth/ps-dispatch)** - For the police dispatch calls  


## Installation Instruction

1. Move "trespassing-alarms" into your resource folder and ensure it
2. Move the "Tresspass_Alarm.ogg" into interact-sound/client/html/sounds
3. Update the "config.lua" to your liking

## Config.lua file
```lua
Config = {}

-- Enable Debugging (Always Alert)
Config.debug = false

Config.zones = {
    [1] = {
        label = "Lucky Plucker", -- Label for your convenience
        zones = { -- Polyzone Coords
            vector2(-68.64, 6268.05),
            vector2(-67.48, 6266.01),
            vector2(-73.08, 6262.27),
            vector2(-74.09, 6264.62),
        },
        minZ = 29.7,
        maxZ = 33.1,
        ignoredJobs = { -- Set jobs that will be ignored
            "police",
            "ambulance"
        },
        -- Alert Configuration
        alertChance = 20,
        alertCooldown = 10, -- Time in seconds
        alarmSound = 'Tresspass_Alarm' -- sound file in interact-sound
    },
    [2] = {
        label = "LS Water & Power", -- Label for your convenience
        zones = { -- Polyzone Coords
            vector2(733.28, 130.44),
            vector2(731.22, 131.09),
            vector2(733.88, 135.59),
            vector2(735.7, 134.5),
        },
        minZ = 79.7,
        maxZ = 83.1,
        ignoredJobs = { -- Set jobs that will be ignored
            "police",
            "ambulance"
        },
        -- Alert Configuration
        alertChance = 20,
        alertCooldown = 10, -- Time in seconds
        alarmSound = 'Tresspass_Alarm' -- sound file in interact-sound
    },
}
```

## Zone setup
```lua
[1] = {
        label = "Lucky Plucker", -- Label for your convenience
        zones = { -- Polyzone Coords
            vector2(-68.64, 6268.05),
            vector2(-67.48, 6266.01),
            vector2(-73.08, 6262.27),
            vector2(-74.09, 6264.62),
        },
        minZ = 29.7,
        maxZ = 33.1,
        ignoredJobs = { -- Set jobs that will be ignored
            "police",
            "ambulance"
        },
        -- Alert Configuration
        alertChance = 20,
        alertCooldown = 10, -- Time in seconds
        alarmSound = 'Tresspass_Alarm' -- sound file in interact-sound
    },
  ```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.1] - 2022-07-10

I'm super excited to announce `tresspassing-alarms` Version 1.0.1

This update adds some more customization to the alert system and the code has been cleaned up

**Feel free to report any bugs with the resource to me on Github or discord Nikki#9699**

### Major changes

- Added `Config.defaultNotificationText` This is the default shown text for player notifications
- Added `Config.zones.playSound` Choose whether to play a sound on alerts or not (Pretty useful for silent alarms)
- Added `Config.zones.soundVolume` Choose the volume of the sound played
- Added `Config.zones.alertPlayer` Choose whether to alert the player or not through a QBCore Notification

### Minor Changes
- Cleaned up the code in the `client.lua` file
- Updated `fxmanifest.lua` file

## [1.0.0] - 2022-07-09

I'm super excited to announce `tresspassing-alarms` Version 1.0.0

- A resource where polyzones are setup to protect an area
- Specific Jobs can be ignored (Changeable in config.lua) (If left empty it will always trigger)
- Chance for alert can be changed per polyzone (Changeable in config.lua)
- Set cooldown to prevent continous trigger (Changeable in config.lua)


trespassing-alarms/client.lua
local QBCore = exports['qb-core']:GetCoreObject()


local PlayerData = QBCore.Functions.GetPlayerData()
local PlayerJob = {}
PlayerData = QBCore.Functions.GetPlayerData()
PlayerJob = QBCore.Functions.GetPlayerData().job

RegisterNetEvent('QBCore:Client:OnJobUpdate', function(JobInfo)
  PlayerData.job = JobInfo
end)


function dump(o)
  if type(o) == 'table' then
     local s = '{ '
     for k,v in pairs(o) do
        if type(k) ~= 'number' then k = '"'..k..'"' end
        s = s .. '['..k..'] = ' .. dump(v) .. ','
     end
     return s .. '} '
  else
     return tostring(o)
  end
end

local function has_value (tab, val)
  for index, value in ipairs(tab) do
      if value == val then
          return true
      end
  end

  return false
end


function randomChance (percent) -- returns true a given percentage of calls
  assert(percent >= 0 and percent <= 100) -- sanity check
  return percent >= math.random(1, 100)   -- 1 succeeds 1%, 50 succeeds 50%,
  -- 100 always succeeds, 0 always fails
end

CreateThread(function()
    for k in pairs(Config.zones) do
      local ZoneTriggered = false
      local ZoneLocation = PolyZone:Create(Config.zones[k].zones, {
        name = Config.zones[k].label,
        minZ = Config.zones[k].minz,
        maxZ = Config.zones[k].maxz,
        debugPoly = false
      })

      ZoneLocation:onPlayerInOut(function(isPointInside)
        if isPointInside then
          inZone = true
        else
          inZone = false
        end

        --Check if in zone
        if inZone then -- User is in zone
          local jobs = Config.zones[k].ignoredJobs
          local PlayerJob = PlayerJob.name
          local alert = false
          local triggertime = ''
          -- Check if job is allowed to be in zone
          if has_value(jobs, PlayerData.job.name) then
            alert = false
          else
            alert = true
          end
          -- Create Alert

          if Config.debug == true then

            -- Check if sound alert gets played
            if Config.zones[k].playSound == true then
              TriggerServerEvent("InteractSound_SV:PlayOnSource", Config.zones[k].alarmSound, Config.zones[k].soundVolume)
            end

            -- Insert wait to prevent overlapping sounds
            Wait(1000)

            -- Check if player gets alerted
            if Config.zones[k].alertPlayer == true then
              -- Get the notification text, otherwise fall back to default
              if Config.zones[k].notificationText == '' then
                Config.zones[k].notificationText = Config.defaultNotificationText
              end
              -- Send notification
              QBCore.Functions.Notify(Config.zones[k].notificationText, "success")
            end
            -- Alert the police
            exports['ps-dispatch']:SuspiciousActivity()
          else
            if alert == true then
              -- Generate chance to trigger alert
              local triggerAlert = randomChance(Config.zones[k].alertChance)
              -- If alert is triggered, and zone cooldown is false
                if triggerAlert == true and ZoneTriggered == false then

                  -- Check if sound alert gets played
                  if Config.zones[k].playSound == true then
                    TriggerServerEvent("InteractSound_SV:PlayOnSource", Config.zones[k].alarmSound, Config.zones[k].soundVolume)
                  end
                  -- Insert wait to prevent overlapping sounds
                  Wait(1000)
                  -- Check if player gets alerted
                  if Config.zones[k].alertPlayer == true then
                    -- Get the notification text, otherwise fall back to default
                    if Config.zones[k].notificationText == '' then
                      Config.zones[k].notificationText = Config.defaultNotificationText
                    end
                    -- Send notification
                    QBCore.Functions.Notify(Config.zones[k].notificationText, "success")
                  end
                   -- Alert the police
                exports['ps-dispatch']:SuspiciousActivity()
                ZoneTriggered = true
              end
              Wait(Config.zones[k].alertCooldown * 1000)
              ZoneTriggered = false
            end
          end
        else

        end



      end)
    end
  end)


trespassing-alarms/config.lua
Config = {}

-- Enable Debugging (Always Alert)
Config.debug = false

Config.defaultNotificationText = 'Shit... the cops have been alerted! Quickly get away!'

Config.zones = {
    [1] = {
        label = "Lucky Plucker", -- Label for your convenience
        zones = { -- Polyzone Coords
            vector2(-68.64, 6268.05),
            vector2(-67.48, 6266.01),
            vector2(-73.08, 6262.27),
            vector2(-74.09, 6264.62),
        },
        minZ = 29.7,
        maxZ = 33.1,
        ignoredJobs = { -- Set jobs that will be ignored
            "police",
            "ambulance"
        },
        -- Alert Configuration
        alertChance = 25, -- Chance to alert the cops
        alertCooldown = 10, -- Time in seconds for cooldown to prevent double trigger
        playSound = true, -- Play sound yes or no
        soundVolume = 0.8, -- Volume of the sound played
        alertPlayer = true, -- Alert player yes or no with a notification
        alarmSound = 'Tresspass_Alarm', -- sound file in interact-sound
        -- Notification Configuration
        notificationText = '' -- Custom text to show, otherwise default is used
    },
    [2] = {
        label = "LS Water & Power", -- Label for your convenience
        zones = { -- Polyzone Coords
            vector2(733.28, 130.44),
            vector2(731.22, 131.09),
            vector2(733.88, 135.59),
            vector2(735.7, 134.5),
        },
        minZ = 79.7,
        maxZ = 83.1,
        ignoredJobs = { -- Set jobs that will be ignored
            "police",
            "ambulance"
        },
        -- Alert Configuration
        alertChance = 25, -- Chance to alert the cops
        alertCooldown = 10, -- Time in seconds for cooldown to prevent double trigger
        playSound = true, -- Play sound yes or no
        soundVolume = 0.8, -- Volume of the sound played
        alertPlayer = true, -- Alert player yes or no with a notification
        alarmSound = 'Tresspass_Alarm', -- sound file in interact-sound
        -- Notification Configuration
        notificationText = 'LS Water & Power called the police...' -- Custom text to show, otherwise default is used
    },
}
